src/matchbox/match_index.py

```diff
@@ -0,0 +1,212 @@
+            rejected by an unknown value.
+          - Included objects should be rejected for everything else than values they are indexed with.
+
+
+    .. note ::
+        Due to relying heavily on dictionaries and sets, MatchIndex indexes only hashable objects by one or more
+        hashable values.
+
+    Example of indexing target objects each by single value:
+
+    +--------+-----------+----------------+
+    | Target | Match     | Indexed values |
+    +========+===========+================+
+    | Ob1    | Excluding | 1              |
+    +--------+-----------+----------------+
+    | Ob2    | Including | 3              |
+    +--------+-----------+----------------+
+    | Ob3    | Excluding | 5              |
+    +--------+-----------+----------------+
+    | Ob4    | Excluding |                |
+    +--------+-----------+----------------+
+    | Ob5    | Including | 1              |
+    +--------+-----------+----------------+
+
+    Will result in an index:
+
+    +---------+------------------+
+    | Value   | Excluded targets |
+    +=========+==================+
+    | 1       | Ob1, Ob2         |
+    +---------+------------------+
+    | 3       | Ob5              |
+    +---------+------------------+
+    | 5       | Ob2, Ob3, Ob5    |
+    +---------+------------------+
+    | Any new | Ob2, Ob5         |
+    +---------+------------------+
+
+
+    Example of indexing target objects each by multiple values:
+
+    +--------+-----------+----------------+
+    | Target | Match     | Indexed values |
+    +========+===========+================+
+    | Ob1    | Excluding | 1, 2           |
+    +--------+-----------+----------------+
+    | Ob2    | Including | 3, 4, 7        |
+    +--------+-----------+----------------+
+    | Ob3    | Excluding | 5, 6           |
+    +--------+-----------+----------------+
+    | Ob4    | Excluding |                |
+    +--------+-----------+----------------+
+    | Ob5    | Including | 1, 7           |
+    +--------+-----------+----------------+
+
+    Should result in this index:
+
+    +-----------+------------------+
+    | Value     | Excluded targets |
+    +===========+==================+
+    | 1         | Ob1, Ob2         |
+    +-----------+------------------+
+    | 2         | Ob1, Ob2, Ob5    |
+    +-----------+------------------+
+    | 3         | Ob5              |
+    +-----------+------------------+
+    | 4         | Ob5              |
+    +-----------+------------------+
+    | 5         | Ob2, Ob3, Ob5    |
+    +-----------+------------------+
+    | 6         | Ob2, Ob3, Ob5    |
+    +-----------+------------------+
+    | 7         |                  |
+    +-----------+------------------+
+    | Any other | Ob2, Ob5         |
+    +-----------+------------------+
+    """
+
+    def __init__(self):
+        """Initialize the index."""
+        self.included_targets = set()
```

Comment:
Things we match with indexes. They are first called that way in the first line of the class docstring. I was trying to find some neutral name, not indicating excluding nor including.